File: framec_tests/cpp/Branch/BranchTest.cpp
#include <cassert>
#include <vector>
#include <string>
#include "branch.h"

class BranchController : public Branch {
public:
    BranchController() : Branch() {}

protected:
    void log_do(const std::string& msg) {
        tape.push_back(msg);
    }
};

void TestSimpleIfBool() {
    BranchController sm;
    sm.A();
    sm.OnBool(true);

    assert(sm.state_info() == "7");
    std::vector<std::string> expected = {"then 1", "then 2"};
    assert(sm.state_info() == "7");
    assert(sm.tape == expected);

    sm = BranchController();
    sm.A();
    sm.OnBool(false);

    assert(sm.state_info() == "8");
    std::vector<std::string> expected1 = {"else 1", "else 2"};
    assert(sm.state_info() == "8");
    assert(sm.tape == expected1);
}

void TestSimpleIfInt() {
    BranchController sm;
    sm.A();
    sm.OnInt(7);
    assert(sm.state_info() == "7");
    std::vector<std::string> expected = {"> 5", "< 10", "== 7"};
    assert(sm.tape == expected);

    sm = BranchController();
    sm.A();
    sm.OnInt(-3);
    assert(sm.state_info() == "8");
    std::vector<std::string> expected1 = {"<= 5", "< 10", "!= 7"};
    assert(sm.tape == expected1);

    sm = BranchController();
    sm.A();
    sm.OnInt(12);
    assert(sm.state_info() == "8");
    std::vector<std::string> expected2 = {"> 5", ">= 10", "!= 7"};
    assert(sm.tape == expected2);
}

void TestNegatedIfBool() {
    BranchController sm;
    sm.B();
    sm.OnBool(true);
    assert(sm.state_info() == "8");
    std::vector<std::string> expected = {"else 1", "else 2"};
    assert(sm.tape == expected);

    sm = BranchController();
    sm.B();
    sm.OnBool(false);
    assert(sm.state_info() == "7");
    std::vector<std::string> expected1 = {"then 1", "then 2"};
    assert(sm.tape == expected1);
}

void TestNegatedIfInt() {
    BranchController sm;
    sm.B();
    sm.OnInt(7);
    assert(sm.state_info() == "7");
    std::vector<std::string> expected = {">= 5", "<= 10", "== 7"};
    assert(sm.tape == expected);

    sm = BranchController();
    sm.B();
    sm.OnInt(5);
    assert(sm.state_info() == "8");
    std::vector<std::string> expected1 = {">= 5", "<= 10", "!= 7"};
    assert(sm.tape == expected1);

    sm = BranchController();
    sm.B();
    sm.OnInt(10);
    assert(sm.state_info() == "8");
    std::vector<std::string> expected2 = {">= 5", "<= 10", "!= 7"};
    assert(sm.tape == expected2);

    sm = BranchController();
    sm.B();
    sm.OnInt(0);
    assert(sm.state_info() == "8");
    std::vector<std::string> expected3 = {"< 5", "<= 10", "!= 7"};
    assert(sm.tape == expected3);

    sm = BranchController();
    sm.B();
    sm.OnInt(100);
    assert(sm.state_info() == "8");
    std::vector<std::string> expected4 = {">= 5", "> 10", "!= 7"};
    assert(sm.tape == expected4);
}

void OperatorPrecedenceTest() {
    BranchController sm;
    sm.C();
    sm.OnInt(0);
    std::vector<std::string> expected = {"then 1", "else 2", "then 3", "then 4"};
    assert(sm.tape == expected);
    sm.tape.clear();
    sm.OnInt(7);
    std::vector<std::string> expected1 = {"else 1", "then 2", "else 3", "then 4"};
    assert(sm.tape == expected1);
    sm.tape.clear();
    sm.OnInt(-3);
    std::vector<std::string> expected2 = {"then 1", "else 2", "else 3", "else 4"};
    assert(sm.tape == expected2);
    sm.tape.clear();
    sm.OnInt(12);
    std::vector<std::string> expected3 = {"else 1", "else 2", "then 3", "else 4"};
    assert(sm.tape == expected3);
    sm.tape.clear();
}

void NestedIfTest() {
    BranchController sm;
    sm.D();
    sm.OnInt(50);
    assert(sm.state_info() == "7");
    std::vector<std::string> expected = {"> 0", "< 100"};
    assert(sm.tape == expected);

    sm = BranchController();
    sm.D();
    sm.OnInt(200);
    assert(sm.state_info() == "4");
    std::vector<std::string> expected1 = {"> 0", ">= 100"};
    assert(sm.tape == expected1);

    sm = BranchController();
    sm.D();
    sm.OnInt(-5);
    assert(sm.state_info() == "4");
    std::vector<std::string> expected2 = {"<= 0", "> -10"};
    assert(sm.tape == expected2);

    sm = BranchController();
    sm.D();
    sm.OnInt(-10);
    assert(sm.state_info() == "8");
    std::vector<std::string> expected3 = {"<= 0", "<= -10"};
    assert(sm.tape == expected3);
}

void TestGuardedTransition() {
    BranchController sm;
    sm.E();
    sm.OnInt(5);
    assert(sm.state_info() == "9");
    std::vector<std::string> expected = {"-> $F3"};
    assert(sm.tape == expected);

    sm = BranchController();
    sm.E();
    sm.OnInt(15);
    assert(sm.state_info() == "8");
    std::vector<std::string> expected1 = {"-> $F2"};
    assert(sm.tape == expected1);

    sm = BranchController();
    sm.E();
    sm.OnInt(115);
    assert(sm.state_info() == "7");
    std::vector<std::string> expected2 = {"-> $F1"};
    assert(sm.tape == expected2);
}

void TestNestedGuardedTransition() {
    BranchController sm;
    sm.F();
    sm.OnInt(5);
    assert(sm.state_info() == "9");
    std::vector<std::string> expected = {"-> $F3"};
    assert(sm.tape == expected);
    sm = BranchController();
    sm.F();
    sm.OnInt(15);
    assert(sm.state_info() == "8");
    std::vector<std::string> expected1 = {"-> $F2"};
    assert(sm.tape == expected1);

    sm = BranchController();
    sm.F();
    sm.OnInt(65);
    assert(sm.state_info() == "9");
    std::vector<std::string> expected2 = {"-> $F3"};
    assert(sm.tape == expected2);

    sm = BranchController();
    sm.F();
    sm.OnInt(115);
    assert(sm.state_info() == "7");
    std::vector<std::string> expected3 = {"-> $F1"};
    assert(sm.tape == expected3);
}

int main() {
    TestSimpleIfBool();
    TestSimpleIfInt();
    TestNegatedIfBool();
    TestNegatedIfInt();
    OperatorPrecedenceTest();
    NestedIfTest();
    TestGuardedTransition();
    TestNestedGuardedTransition();

    return 0;
}

File: framec_tests/java/Branch/Branch_test.java
package framec_tests.java.Branch;

import static org.junit.Assert.assertEquals;

import java.util.ArrayList;

import org.junit.Test;


class BranchController extends Branch {
  BranchController() {
    super();
  }
  protected void log_do(String msg) {
    this.tape.add(msg);
  }
}
public class Branch_test {
    
@Test
public void TestSimpleIfBool(){
    BranchController sm = new BranchController();
    sm.A();
    sm.OnBool(true);

    assertEquals("7", sm.state_info());
    ArrayList<String> expected = new ArrayList<String>();
    expected.add("then 1");
    expected.add("then 2");

    assertEquals(sm.state_info(), "7");
    assertEquals(sm.tape, expected);
    sm = new BranchController();
    sm.A();
    sm.OnBool(false);

    assertEquals("8" ,sm.state_info());
    ArrayList<String> expected1 = new ArrayList<String>();
    expected1.add("else 1");
    expected1.add("else 2");
    assertEquals(sm.state_info(), "8");
    assertEquals(expected1, sm.tape);
  }

  @Test
  public void TestSimpleIfInit(){
    BranchController sm = new BranchController();
    sm.A();
    sm.OnInt(7);
    assertEquals(sm.state_info(), "7" );
    ArrayList<String> arr = new ArrayList<String>();
    arr.add("> 5");
    arr.add("< 10");
    arr.add("== 7");
    assertEquals(sm.tape, arr);
    sm = new BranchController();
    sm.A();
    sm.OnInt(-3);
    assertEquals(sm.state_info(), "8");
    ArrayList<String> arr1 = new ArrayList<String>();
    arr1.add("<= 5");
    arr1.add("< 10");
    arr1.add("!= 7");
    assertEquals(sm.tape, arr1);
    sm = new BranchController();
    sm.A();
    sm.OnInt(12);
    assertEquals(sm.state_info(), "8");
    ArrayList<String> arr2 = new ArrayList<String>();
    arr2.add("> 5");
    arr2.add(">= 10");
    arr2.add("!= 7");
    assertEquals(sm.tape, arr2);

  }

  @Test
  public void TestNegatedIfBool(){
    BranchController sm = new BranchController();
    sm.B();
    sm.OnBool(true);
    assertEquals(sm.state_info(), "8");
    ArrayList<String> expected  = new ArrayList<String>();
    expected.add("else 1");
    expected.add("else 2");
    assertEquals(sm.tape, expected);
    sm = new BranchController();
    sm.B();
    sm.OnBool(false);
    assertEquals(sm.state_info(), "7");
    ArrayList<String> expected1 = new ArrayList<String>();
    expected1.add("then 1");
    expected1.add("then 2");
    assertEquals(sm.tape, expected1);
  }

  @Test
  public void TestNegatedIfInt(){
    BranchController sm = new BranchController();
    sm.B();
    sm.OnInt(7);
    assertEquals(sm.state_info(), "7");
    ArrayList<String> expected = new ArrayList<>();
    expected.add(">= 5");
    expected.add("<= 10");
    expected.add("== 7");
    assertEquals(sm.tape, expected);

    sm = new BranchController();
    sm.B();
    sm.OnInt(5);
    assertEquals(sm.state_info(), "8");
    ArrayList<String> expected1 = new ArrayList<String>();
    expected1.add(">= 5");
    expected1.add("<= 10");
    expected1.add("!= 7");
    assertEquals(sm.tape, expected1);

    sm = new BranchController();
    sm.B();
    sm.OnInt(10);
    assertEquals(sm.state_info(), "8");
    ArrayList<String> expected2 = new ArrayList<String>();
    expected2.add(">= 5");
    expected2.add("<= 10");
    expected2.add("!= 7");
    assertEquals(sm.tape, expected2);

    sm = new BranchController();
    sm.B();
    sm.OnInt(0);
    assertEquals(sm.state_info(), "8");
    ArrayList<String> expected3 = new ArrayList<String>();
    expected3.add("< 5");
    expected3.add("<= 10");
    expected3.add("!= 7");
    assertEquals(sm.tape, expected3);

    sm = new BranchController();
    sm.B();
    sm.OnInt(100);
    assertEquals(sm.state_info(), "8");
    ArrayList<String> expected4 = new ArrayList<String>();
    expected4.add(">= 5");
    expected4.add("> 10");
    expected4.add("!= 7");
    assertEquals(sm.tape, expected4);
  }

  @Test
  public void TestOperatorPrecedence(){
    BranchController sm = new BranchController();
    sm.C();
    sm.OnInt(0);
    ArrayList<String> expected = new ArrayList<String>();
    expected.add("then 1");
    expected.add("else 2");
    expected.add("then 3");
    expected.add("then 4");
    assertEquals(expected, sm.tape);
    sm.tape = new ArrayList<>();
    sm.OnInt(7);
    ArrayList<String> expected1 = new ArrayList<String>();
    expected1.add("else 1");
    expected1.add("then 2");
    expected1.add("else 3");
    expected1.add("then 4");
    assertEquals(expected1, sm.tape);
    sm.tape = new ArrayList<>();
    sm.OnInt(-3);
    ArrayList<String> expected2 = new ArrayList<String>();
    expected2.add("then 1");
    expected2.add("else 2");
    expected2.add("else 3");
    expected2.add("else 4");
    assertEquals(expected2, sm.tape);
    sm.tape = new ArrayList<>();
    sm.OnInt(12);
    ArrayList<String> expected3 = new ArrayList<String>();
    expected3.add("else 1");
    expected3.add("else 2");
    expected3.add("then 3");
    expected3.add("else 4");
    assertEquals(expected3, sm.tape);
    sm.tape = new ArrayList<>();

  }

  @Test
  public void TestNestedIf(){
    BranchController sm = new BranchController();
    sm.D();
    sm.OnInt(50);
    assertEquals(sm.state_info(), "7");
    ArrayList<String> expected = new ArrayList<String>();
    expected.add("> 0");
    expected.add("< 100");
    assertEquals(sm.tape, expected);
    sm = new BranchController();
    sm.D();
    sm.OnInt(200);
    assertEquals(sm.state_info(), "4");
    ArrayList<String> expected1 = new ArrayList<String>();
    expected1.add("> 0");
    expected1.add(">= 100");
    assertEquals(sm.tape, expected1);
    sm = new BranchController();
    sm.D();
    sm.OnInt(-5);
    assertEquals(sm.state_info(), "4");
    ArrayList<String> expected2 = new ArrayList<String>();
    expected2.add("<= 0");
    expected2.add("> -10");
    assertEquals(sm.tape, expected2);
    sm = new BranchController();
    sm.D();
    sm.OnInt(-10);
    assertEquals(sm.state_info(), "8");
    ArrayList<String> expected3 = new ArrayList<String>();
    expected3.add("<= 0");
    expected3.add("<= -10");
    assertEquals(sm.tape, expected3);
    sm = new BranchController();
    
  }
  @Test
  public void TestGuardedTransition(){
    BranchController sm = new BranchController();
    sm.E();
    sm.OnInt(5);
    assertEquals(sm.state_info(), "9");
    ArrayList<String> expected = new ArrayList<String>();
    expected.add("-> $F3");
    assertEquals(sm.tape, expected);
    sm = new BranchController();
    sm.E();
    sm.OnInt(15);
    assertEquals(sm.state_info(), "8");
    ArrayList<String> expected1 = new ArrayList<String>();
    expected1.add("-> $F2");
    assertEquals(sm.tape, expected1);
    sm = new BranchController();
    sm.E();
    sm.OnInt(115);
    assertEquals(sm.state_info(), "7");
    ArrayList<String> expected2 = new ArrayList<>();
    expected2.add("-> $F1");
    assertEquals(sm.tape, expected2);
  }
  /*
	Test that a transition guarded by a nested conditional expression

triggers an early return from the handler, but this return doesn't

	apply to non-transitioned branches.
*/
  @Test
  public void TestNestedGuardedTransition(){
    BranchController sm = new BranchController();
    sm.F();
    sm.OnInt(5);
    assertEquals(sm.state_info(), "9");
    ArrayList<String> expected = new ArrayList<String>();
    expected.add("-> $F3");
    assertEquals(sm.tape, expected);
    sm = new BranchController();
    sm.tape = new ArrayList<>();
    sm.F();
    sm.OnInt(15);
    assertEquals(sm.state_info(), "8");
    ArrayList<String> expected1 = new ArrayList<String>();
    expected1.add("-> $F2");
    assertEquals(sm.tape, expected1);
    sm = new BranchController();
    sm.F();
    sm.OnInt(65);
    assertEquals(sm.state_info(), "9");
    ArrayList<String> expected2 = new ArrayList<String>();
    expected2.add("-> $F3");
    assertEquals(sm.tape, expected2);
    sm = new BranchController();
    sm.F();
    sm.OnInt(115);
    assertEquals(sm.state_info(), "7");
    ArrayList<String> expected3 = new ArrayList<String>();
    expected3.add("-> $F1");
    assertEquals(sm.tape, expected3);
    
  }
}